### PR TITLE
DBZ-2654 Edit Customizing Topic Auto-Creation

### PR DESCRIPTION
**Pull request by roldanbob**
Jira [DBZ-2654](https://issues.redhat.com/browse/DBZ-2654)

**WIP**: Pending meeting with @rk3rn3r to discuss some aspects of the material.

I tested the file as-is in a local Antora build, as well as in a local downstream build.

### CODE REVIEW

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 1]
```diff
@@ -13,19 +13,29 @@
...
+// ModuleID: configuring-automatic-topic-creation-in-kafka-2-6
+// Title: Configuring automatic topic creation in Kafka 2.6 and later
+[id="enabling-automatic-topic-creation"]
 == Set up Kafka Connect
 
-Kafka Connect since Kafka 2.6.0 comes with topic creation enabled:
+In Kafka 2.6.0, the introduction of a new property provides greater flexibility in configuring automatic topic creation.
+You set the `topic.creation.enable` property in Kafka Connect to enable or disable automatic topic creation.
+When automatic topic creation is enabled by this property, you can define multiple configuration profiles and apply them to groups of topics based on topic names.
+The default value for the property is `true`, enabling automatic topic creation, as shown in the following example:
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:38]: So I think this is confusing, as there's two different concepts at play here:

* auto-topic creation by the broker; this has existed before, and means the broker will automatically create a missing topic when receiving a message for a non-existent topic; this is generally not recommended to use as you'll end up with only a single setting for all those auto-created topics
* topic creation via Kafka Connect: this is new, and here Kafka Connect does create any missing topics, based on the connector configuration; i.e. topics are _not_ auto-created by the broker

The two concepts should not be confused in the docs.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 2]
```diff
@@ -35,18 +45,19 @@ topic.creation.enable = true
...
 ifdef::product[]
-[NOTE]
-====
-If you don't want to allow automatic topic creation by connectors you can set this value to `false`
-in the Kafka Connect CRD:
-====
+
+.Procedure
+
+* To prevent connectors from creating topics automatically, set the value of `topic.creation.enable` to `false`
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:58]: It's not exactly connectors but Kafka Connect which creates topics.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 3]
```diff
@@ -63,32 +74,64 @@ spec:
 ----
 endif::product[]
 
+[NOTE]
+====
+Automatic topic creation requires the `replication.factor` and `partitions` properties to be set for at least the `default` topic creation group.
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:79]: See above, I think we should not name this "automatic topic creation". Would be interesting to see how it's described in the Kafka KIP.

**Review by gunnarmorling — COMMENTED** [review]
A few comments inline. Leaving it to @rk3rn3r for further remarks. Thanks!

**Comment by gunnarmorling** [comment]
@roldanbob, @rk3rn3r, this is still evolving, right?

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 3 above)
Comment (roldanbob) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:79]: Topic auto-creation = topic automatic creation = automatic topic creation, doesn't it?  I'm not sure what is your objection. Do you think that using the same terminology to refer to Kafka Connect topic creation and broker topic creation is too confusing? 

In the KIP we see the following phrase: _"Kafka Connect should instead be able to create the topics automatically for source connectors ..."_ Elsewhere there's a reference to _"creation of new topics during runtime"_ and _"topic creation during runtime."_

We could instead refer to _"runtime topic creation"_. WDYT?

**Comment by roldanbob** [comment]
> @roldanbob, @rk3rn3r, this is still evolving, right?

Yes, I've still got some updates to incorporate from my meeting with @rk3rn3r, and I'm waiting for him to review the current draft.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 1 above)
Comment (roldanbob) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:38]: Yes, I'll do a better job of clarifying the distinction.

**Comment by roldanbob** [comment]
Updated draft. I think this is improved/ready and addresses input with the exception of the question about terminology (automatic topic creation).  Let me know whether you think that the terminology issue is a blocker.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 4]
```diff
@@ -13,19 +13,69 @@
...
-{prodname} automatically creates *internal* topics for offsets, connector status, config
-storage and history topics. The destination topics for the captured tables will be
-automatically created with a default config by the Kafka brokers when
-`auto.create.topics.enable` is set to `true`.{empty} +
-When topic creation is disabled on the brokers, for example in production environments,
-or when the topics need a different configuration then these topics have to be created
-upfront either automated in a custom deployment process or manually until Kafka Connect 2.6.
+Beginning with Kafka 2.6.0, you can set the Kafka Connect configuration to permit automatic creation of topics.
+If you enable automatic topic creation,
+when Debezium emits a change event record, Kafka Connect ingests the record into Kafka and writes it to a topic.
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:18]: This part of the sentence is always true, also without the condition on the line before.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 5]
```diff
@@ -13,19 +13,69 @@
...
+Topic are named based on the pattern `server.schema.table`, for example, `dbserver`.myschema.inventory`.
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:19]: By default, topics are named...

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 6]
```diff
@@ -13,19 +13,69 @@
...
+
+* In the broker configuration, set the value of `auto.create.topics.enable` to `false`.
+
+[NOTE]
+====
+In environments that run Kafka version 2.6.0 or later, the Kafka Connect property to enable automatic topic creation is independent of the broker property.
+The Kafka Connect configuration takes precedence over the broker configuration.
+Disabling topic creation at the broker has no effect on whether Kafka Connect can create topics.
+However, if you configure Kafka Connect to create topics, and you retain the default broker configuration,
+the broker creates topics when none of the topic creation groups in the Kafka Connect configuration apply,
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:63]:
```suggestion
the broker creates topics when none of the topic creation groups in the Kafka Connect configuration apply.
```

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 7]
```diff
@@ -13,19 +13,69 @@
...
+// ModuleID: configuring-automatic-topic-creation-in-kafka-connect
+// Title: Configuring automatic topic creation in Kafka Connect
+[id="enabling-automatic-topic-creation"]
 == Set up Kafka Connect
 
-Kafka Connect since Kafka 2.6.0 comes with topic creation enabled:
+In Kafka 2.6 and greater, you can set a Kafka Connect property that specifies whether Kafka Connect can create topics that a {prodname} connector writes if those topics do not already exist.
+The `topic.creation.enable` property in Kafka Connect enables or disables automatic topic creation.
+The setting for the `topic.creation.enable` property applies to all workers in the Connect cluster.
+The default value for the property is `true`, enabling automatic topic creation, as shown in the following example:
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:75]: Where is this  example?

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 8]
```diff
@@ -35,18 +85,19 @@ topic.creation.enable = true
 ifdef::community[]
 [NOTE]
 ====
-If you don't want to allow automatic topic creation by connectors you can set this value to `false`
-in the Kafka Connect config (_connect-distributed.properties_ file or via environment variable
+If you don't want to allow automatic topic creation by connectors, set the value of `topic.creation.enable` to `false`
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:88]: "by Kafka Connect"

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 3 above)
Comment (gunnarmorling) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:119]: Perhaps it needs qualifying? "Kafka Connect automatic topic creation"? Or "Automatic topic creation driven by Kafka Connect"?

> Do you think that using the same terminology to refer to Kafka Connect topic creation and broker topic creation is too confusing?

Yes, that's my concern. But perhaps it's not a problem if things become clear from the context.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 9]
```diff
@@ -63,32 +114,77 @@ spec:
...
+// Title: Configuration of automatically created topics
+[id="configuration"]
 == Configuration
 
-Topic auto-creation is based on groups. Every custom group has an `include` and an
-`exclude` property which are comma-separated lists of regular expressions matching
-topic names that should be included or excluded.
+For Kafka Connect to create topics automatically, it requires information from the source connector about the configuration properties to apply when creating topics.
+You define the properties that control topic creation in the configuration for each {prodname} connector.
+As Kafka Connect create topics for event records that a connector emits, the resulting topics obtain their configuration from the applicable group.
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:131]:
```suggestion
As Kafka Connect creates topics for change event records that a connector emits, the resulting topics obtain their configuration from the applicable group.
```

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 10]
```diff
@@ -63,32 +114,77 @@ spec:
...
+[id="default-topic-creation-group-configuration"]
 === Default group configuration
 
-The default config can be passed in the connector config JSON like:
+You must create a default topic creation group and define a configuration for it.
+The configuration for the default topic creation group is applied to any topics whose names do not match the `include` list pattern of a custom topic creation group.
+
+.Procedure
 
+* To define properties for the `topic.creation.default` group, add them to the connector configuration JSON, as shown in the following example:
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:186]: For downstream, this shouldn't be JSON but rather the AMQ Streams CR.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 11]
```diff
@@ -179,43 +282,50 @@ the group name. This will look like that in your connector JSON:
...
+[id="registering-custom-topic-creation-groups"]
 === Registering custom groups
 
-Finally, we need to register the two defined custom groups `inventory` and `applicationlogs` with
-the `topic.creation.groups` property:
+After you specify the configuration for any custom topic creation groups, register the groups.
+
+.Procedure
 
+* Register custom groups by adding the `topic.creation.groups` property to the connector JSON, and specifying a comma-separated list of groups.
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:325]: JSON should only be used upstream. Same comment below and potentially elsewhere in this doc.

**Review by gunnarmorling — COMMENTED** [review]
@roldanbob, some more comments inline. Most notably, we should use the AMQ Streams CR instead of JSON config for downstream docs.

@rk3rn3r, any further comments from your side?

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 10 above)
Comment (roldanbob) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:186]: > AMQ Streams CR.

I think that what you mean is that on Streams, you must define the properties for the default group in the CR _for the Debezium connector_? Is that correct?

I'd like to provide an example, but I'm having trouble coming up with one that I can model. I believe that you would specify the properties in `spec.config`, yes? So something like this:

```
spec:
...

   config:
...
     topic.creation.default.replication.factor: 3 
     topic.creation.default.partitions: 10  
     topic.creation.default.cleanup.policy: compact  
     topic.creation.default.compression.type: lz4  
```

Comment (roldanbob) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:186]: Added a conditionalized example for downstream that shows the configuration as an excerpt from a connector custom resource.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 12]
```diff
@@ -13,40 +13,96 @@
...
+The setting for the `topic.creation.enable` property applies to all workers in the Connect cluster.
+
+Kafka Connect automatic topic creation requires you to define the configuration properties that Kafka Connect applies when creating topics.
+You specify topic configuration properties in the {prodname} connector configuration by defining topic groups, and then specifying the properties to apply to each group.
+The connector configuration defines a default topic creation group, and, optionally, one or more custom topic creation groups.
+Custom topic creation groups use lists of topic name patterns to specify the topics to which the group's settings apply.
+For details about how Kafka Connect matches topics to topic creation groups, see xref:topic-creation-groups[].
+For more information about how configuration properties are assigned to groups, see xref:topic-creation-group-configuration-properties[Topic creation group configuration properties].
+
+By default, topics that Kafka Connect creates are named based on the pattern `server.schema.table`, for example, `dbserver`.myschema.inventory`.
```
Comment (rk3rn3r) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:88]: not enough or too many ` `` ` signs?

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 13]
```diff
@@ -130,32 +276,38 @@ defines the log retention policy.
...
+The following example shows sample configurations for the `inventory` and `applicationlogs` custom topic creation groups:
++
 [source,options="nowrap",json]
 ----
 {
     ...
 
     //<1>
     "topic.creation.inventory.include": "dbserver1\\.inventory\\.*",  //<2>
+    "topic.creation.inventory.replication.factor": 3,
```
Comment (rk3rn3r) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:310]: I think I left out that `replication.factor` property on purpose to showcase the above note, that `Custom groups fall back to the `default` group settings for the required "replication.factor" and "partitions" properties`

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 14]
```diff
@@ -172,62 +324,134 @@ the group name. This will look like that in your connector JSON:
...
+metadata:
+  name: inventory-connector
+...
+spec:
+...
+
+   config:
+... //<1>
+     topic.creation.inventory.include: "dbserver1\\.inventory\\.*"  //<2>
+     "topic.creation.inventory.replication.factor": 3
```
Comment (rk3rn3r) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:347]: remove the double quotes here

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 15]
```diff
@@ -253,7 +477,41 @@ A complete connector JSON config will look like that:
...
+
+   config:
+...
+    topic.creation.default.replication.factor: 3,
+    topic.creation.default.partitions: 10,
+    topic.creation.default.cleanup.policy: compact
+    topic.creation.default.compression.type": lz4
+    topic.creation.groups: inventory,applicationlogs
+    topic.creation.inventory.include: dbserver1\\.inventory\\.*
+    topic.creation.inventory.replication.factor: 3
```
Comment (rk3rn3r) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:501]: again, not needed to repeat the default group value here, as long there's no foreseeable reason to update the default group's value in the future and keep that value for this group.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 16]
```diff
@@ -130,32 +276,38 @@ defines the log retention policy.
...
+[id="custom-topic-creation-group-configuration"]
 === Custom group configuration
 
-You can specify multiple groups. Similar to the `default` group you group properties together by
-the group name. This will look like that in your connector JSON:
+You can define multiple custom topic groups, each with its own configuration.
 
+.Procedure
+ifdef::community[]
+* To define a custom topic group, add a `topic.creation._<group_name>_.include` property to the connector JSON, and
```
[the remark this answers is not in the document]
Comment (roldanbob) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:298]: > is it possible in asciidoc to have `_` as italic styling inside a `` `` `` block?

Yes, our standard downstream for styling replaceable content in a code example is an italicized string of lower case text enclosed in angle brackets.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 12 above)
Comment (roldanbob) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:88]: Oops. Yes. Thanks, for catching that ... one extra there. Fixed it.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 13 above)
Comment (roldanbob) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:310]: Oh, ok. Yup, I made a mistake when adding the CR example, and inadvertently overwrote the original JSON example. Bad copy/paste error when I went to restore it, which I then propagated to the completed example later. :(

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 14 above)
Comment (roldanbob) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:347]: Ah, thanks, I missed missed seeing the extra quotes. :+1: Fixed it.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 15 above)
Comment (roldanbob) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:501]: Thanks for your close review. Sorry about the silly mistakes. :(

**Review by gunnarmorling — APPROVED** [review]
LGTM now. @rk3rn3r, can you take one last look and merge if it's good for you, too? Thanks to you and @roldanbob!

**Pull request merged**